edb/edgeql/compiler/casts.py

```diff
@@ -157,9 +157,18 @@ def compile_cast(
             or new_stype.issubclass(ctx.env.schema, json_t)
         )
     ):
-        return _cast_array_literal(
+        cast_element = ('array', None)
+        if ctx.collection_cast_path is not None:
+            ctx.collection_cast_path.append(cast_element)
```

Comment:
So the idea is that if the compile fails, then the pops don't happen and the message gets collected at the end?

That's a little subtle, though might still be fine.

Are there too many base cases for this error to inject the message closer to the initial sites where it is raised?

Reply:
(OK yeah, and that pattern is properly documented too)

Reply:
There "should" always be a base TypeCast operation which handles all this.